tracer_metrics.go

```diff
@@ -0,0 +1,369 @@
+package libp2pquic
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/lucas-clemente/quic-go/logging"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	bytesTransferred *prometheus.CounterVec
+	newConns         *prometheus.CounterVec
+	closedConns      *prometheus.CounterVec
+	sentPackets      *prometheus.CounterVec
+	rcvdPackets      *prometheus.CounterVec
+	bufferedPackets  *prometheus.CounterVec
+	droppedPackets   *prometheus.CounterVec
+	lostPackets      *prometheus.CounterVec
+	connErrors       *prometheus.CounterVec
+)
+
+type aggregatingCollector struct {
+	mutex sync.Mutex
+
+	conns         map[string] /* conn ID */ *metricsConnTracer
+	rtts          prometheus.Histogram
+	connDurations prometheus.Histogram
+}
+
+func newAggregatingCollector() *aggregatingCollector {
+	return &aggregatingCollector{
+		conns: make(map[string]*metricsConnTracer),
+		rtts: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "quic_smoothed_rtt",
+			Help:    "Smoothed RTT",
+			Buckets: prometheus.ExponentialBuckets(0.001, 1.25, 40), // 1ms to ~6000ms
+		}),
+		connDurations: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "quic_connection_duration",
+			Help:    "Connection Duration",
+			Buckets: prometheus.ExponentialBuckets(1, 1.5, 40), // 1s to ~12 weeks
+		}),
+	}
+}
+
+var _ prometheus.Collector = &aggregatingCollector{}
+
+func (c *aggregatingCollector) Describe(descs chan<- *prometheus.Desc) {
+	descs <- c.rtts.Desc()
+	descs <- c.connDurations.Desc()
+}
+
+func (c *aggregatingCollector) Collect(metrics chan<- prometheus.Metric) {
+	now := time.Now()
+	c.mutex.Lock()
+	for _, conn := range c.conns {
+		if rtt, valid := conn.getSmoothedRTT(); valid {
+			c.rtts.Observe(rtt.Seconds())
+		}
+		c.connDurations.Observe(now.Sub(conn.startTime).Seconds())
+	}
+	c.mutex.Unlock()
+	metrics <- c.rtts
+	metrics <- c.connDurations
+}
+
+func (c *aggregatingCollector) AddConn(id string, t *metricsConnTracer) {
+	c.mutex.Lock()
+	c.conns[id] = t
+	c.mutex.Unlock()
+}
+
+func (c *aggregatingCollector) RemoveConn(id string) {
+	c.mutex.Lock()
+	delete(c.conns, id)
+	c.mutex.Unlock()
+}
+
+var collector *aggregatingCollector
+
+func init() {
+	const (
+		direction = "direction"
+		encLevel  = "encryption_level"
+	)
+
+	closedConns = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_closed_connections",
```

Comment:
```suggestion
			Name: "quic_connections_closed_total",
```

Small nit: There are a couple of [metric naming conventions](https://prometheus.io/docs/practices/naming/). One of them being that unit-less counters end with `_total`. Same applies to the unit-less metrics below.

Reply:
Thanks, I renamed most of the metrics.